### ISSUE TITLE
WARNING: unstable losses; when training on custom dataset

### ISSUE THREAD

**Issue by MaddyThakker** [divyam02/dafrcnn-pytorch#11]
Thanks for the PyTorch implementation of the Domain Adaptive F-RCNN. While training the resnet-101 network on the custom dataset, after training for 1-2 epochs, I start getting a Warning: Unstable Losses. 
If I let it train even after getting the warning, after a couple of epochs, I get an error - "encountered nan".  
I have tried with batch-size values of 1 and 2. My source domain images are (5120, 5120, 3) dimension and target domain images are (720, 1280, 3) dimension. I have around 250 annotated source images and 25 target images.

**Comment by MaddyThakker** [comment]
To add more to the issue, this is the output I get. 

Command - `CUDA_VISIBLE_DEVICES=0 python new_trainval.py --net res101 --cuda --nw 2 --epochs 20 --use_tfb --checkpoint_interval 5 --bs 1 --disp_interval 10 --start_epoch 0 --save_dir 'models' --checksession 2`

```
Starting factory.py
Called with args:
Namespace(batch_size=1, checkepoch=1, checkpoint=0, checkpoint_interval=5, checksession=2, class_agnostic=False, cuda=True, dataset='city', disp_interval=10, large_scale=False, lr=0.001, lr_decay_gamma=0.1, lr_decay_step=50000, mGPUs=False, max_epochs=20, net='res101', num_workers=2, optimizer='sgd', resume=False, save_dir='models', session=1, start_epoch=0, stop_iter=70000, target_dataset='fcity', use_tfboard=True)
the source dataset is indeed city
the target dataset is indeed fcity
Using config:
{'ANCHOR_RATIOS': [0.5, 1, 2],
 'ANCHOR_SCALES': [8, 16, 32],
 'CROP_RESIZE_WITH_MAX_POOL': False,
 'CUDA': False,
 'DATA_DIR': '/content/drive/My '
             'Drive/BTP/DA-FRCNN/DA-FRCNN-Pytorch/dafrcnn-pytorch/data',
 'DEDUP_BOXES': 0.0625,
 'EPS': 1e-14,
 'EXP_DIR': 'res101',
 'FEAT_STRIDE': [16],
 'GPU_ID': 0,
 'MATLAB': 'matlab',
 'MAX_NUM_GT_BOXES': 500,
 'MOBILENET': {'DEPTH_MULTIPLIER': 1.0,
               'FIXED_LAYERS': 5,
               'REGU_DEPTH': False,
               'WEIGHT_DECAY': 4e-05},
 'PIXEL_MEANS': array([[[102.9801, 115.9465, 122.7717]]]),
 'POOLING_MODE': 'align',
 'POOLING_SIZE': 7,
 'RESNET': {'FIXED_BLOCKS': 1, 'MAX_POOL': False},
 'RNG_SEED': 3,
 'ROOT_DIR': '/content/drive/My '
             'Drive/BTP/DA-FRCNN/DA-FRCNN-Pytorch/dafrcnn-pytorch',
 'TEST': {'BBOX_REG': True,
          'HAS_RPN': True,
          'MAX_SIZE': 1000,
          'MODE': 'nms',
          'NMS': 0.3,
          'PROPOSAL_METHOD': 'gt',
          'RPN_MIN_SIZE': 16,
          'RPN_NMS_THRESH': 0.7,
          'RPN_POST_NMS_TOP_N': 300,
          'RPN_PRE_NMS_TOP_N': 6000,
          'RPN_TOP_N': 5000,
          'SCALES': [600],
          'SVM': False},
 'TRAIN': {'ASPECT_GROUPING': False,
           'BATCH_SIZE': 128,
           'BBOX_INSIDE_WEIGHTS': [1.0, 1.0, 1.0, 1.0],
           'BBOX_NORMALIZE_MEANS': [0.0, 0.0, 0.0, 0.0],
           'BBOX_NORMALIZE_STDS': [0.1, 0.1, 0.2, 0.2],
           'BBOX_NORMALIZE_TARGETS': True,
           'BBOX_NORMALIZE_TARGETS_PRECOMPUTED': True,
           'BBOX_REG': True,
           'BBOX_THRESH': 0.5,
           'BG_THRESH_HI': 0.5,
           'BG_THRESH_LO': 0.0,
           'BIAS_DECAY': False,
           'BN_TRAIN': False,
           'DISPLAY': 20,
           'DOUBLE_BIAS': False,
           'FG_FRACTION': 0.25,
           'FG_THRESH': 0.5,
           'GAMMA': 0.1,
           'HAS_RPN': True,
           'IMS_PER_BATCH': 1,
           'LEARNING_RATE': 0.001,
           'MAX_SIZE': 1000,
           'MOMENTUM': 0.9,
           'PROPOSAL_METHOD': 'gt',
           'RPN_BATCHSIZE': 256,
           'RPN_BBOX_INSIDE_WEIGHTS': [1.0, 1.0, 1.0, 1.0],
           'RPN_CLOBBER_POSITIVES': False,
           'RPN_FG_FRACTION': 0.5,
           'RPN_MIN_SIZE': 8,
           'RPN_NEGATIVE_OVERLAP': 0.3,
           'RPN_NMS_THRESH': 0.7,
           'RPN_POSITIVE_OVERLAP': 0.7,
           'RPN_POSITIVE_WEIGHT': -1.0,
           'RPN_POST_NMS_TOP_N': 2000,
           'RPN_PRE_NMS_TOP_N': 12000,
           'SCALES': [600],
           'SNAPSHOT_ITERS': 5000,
           'SNAPSHOT_KEPT': 3,
           'SNAPSHOT_PREFIX': 'res101_faster_rcnn',
           'STEPSIZE': [30000],
           'SUMMARY_INTERVAL': 180,
           'TRIM_HEIGHT': 600,
           'TRIM_WIDTH': 600,
           'TRUNCATED': False,
           'USE_ALL_GT': True,
           'USE_FLIPPED': True,
           'USE_GT': False,
           'WEIGHT_DECAY': 0.0001},
 'USE_GPU_NMS': True}
Dataset:
 city_2007_trainval
initialized imdb...
('__background__', 'bee')
sample image from dataset: frame_30fps_006465
Loaded dataset `voc_2007_trainval` for training
Set proposal method: gt

[session 1][epoch  0][iter    0/ 710][steps    1] loss: 4.2746, lr: 1.00e-03
			fg/bg=(32/96), time cost: 6.531387
			rpn_cls: 0.7153, rpn_box: 2.7030, rcnn_cls: 0.5410, rcnn_box 0.0019
			src_img_loss: 0.7332, src_inst_loss: 0.7238, src_consistency_loss: 0.0362
			tar_img_loss: 0.8782, tar_inst_loss: 0.6731, tar_consistency_loss: 0.0892

[session 1][epoch  0][iter   10/ 710][steps   11] loss: 1.9873, lr: 1.00e-03
			fg/bg=(32/96), time cost: 19.899629
			rpn_cls: 0.2183, rpn_box: 0.6946, rcnn_cls: 0.4251, rcnn_box 0.0026
			src_img_loss: 0.7371, src_inst_loss: 0.7280, src_consistency_loss: 0.0274
			tar_img_loss: 0.6477, tar_inst_loss: 0.6630, tar_consistency_loss: 0.0365
.....
.....
.....
[session 1][epoch  0][iter  590/ 710][steps  591] loss: 0.8258, lr: 1.00e-03
			fg/bg=(32/96), time cost: 19.971207
			rpn_cls: 0.0044, rpn_box: 0.1641, rcnn_cls: 0.3102, rcnn_box 0.0491
			src_img_loss: 0.6900, src_inst_loss: 1.4980, src_consistency_loss: 0.2181
			tar_img_loss: 0.5802, tar_inst_loss: 0.2300, tar_consistency_loss: 0.3222

[session 1][epoch  0][iter  600/ 710][steps  601] loss: 0.9722, lr: 1.00e-03
			fg/bg=(32/96), time cost: 20.476809
			rpn_cls: 0.0064, rpn_box: 0.1286, rcnn_cls: 0.3549, rcnn_box 0.1176
			src_img_loss: 0.4183, src_inst_loss: 2.4073, src_consistency_loss: 0.3390
			tar_img_loss: 2.7348, tar_inst_loss: 0.1969, tar_consistency_loss: 0.8324

WARNING: unstable losses!
tensor([[[[ 6.2432e+01,  6.4187e+01,  6.2653e+01,  ...,  8.2979e+00,
            7.0199e+00,  5.2533e+00],
          [ 6.4320e+01,  6.5320e+01,  6.6737e+01,  ...,  2.5986e+01,
            2.2010e+01,  2.7090e+01],
          [ 6.7214e+01,  6.8553e+01,  6.4353e+01,  ...,  3.0881e+01,
            2.7670e+01,  2.9592e+01],
          ...,
          [ 4.3298e+01,  4.3453e+01,  4.6353e+01,  ..., -8.2147e+01,
           -8.1897e+01, -8.1147e+01],
          [ 4.0550e+01,  4.3210e+01,  4.8553e+01,  ..., -8.3797e+01,
           -8.8081e+01, -8.1980e+01],
          [ 4.6771e+01,  4.5996e+01,  4.5398e+01,  ..., -8.4147e+01,
           -8.5747e+01, -8.0980e+01]],

         [[ 4.9466e+01,  5.1220e+01,  4.9687e+01,  ..., -4.6685e+00,
           -5.9465e+00, -7.7131e+00],
          [ 5.1354e+01,  5.2354e+01,  5.3770e+01,  ...,  1.3020e+01,
            9.0434e+00,  1.4124e+01],
          [ 5.4248e+01,  5.5587e+01,  5.1387e+01,  ...,  1.7915e+01,
            1.4704e+01,  1.6626e+01],
          ...,
          [ 3.0332e+01,  3.0487e+01,  3.3387e+01,  ..., -9.5113e+01,
           -9.4863e+01, -9.4113e+01],
          [ 2.7583e+01,  3.0244e+01,  3.5587e+01,  ..., -9.6763e+01,
           -1.0105e+02, -9.4947e+01],
          [ 3.3805e+01,  3.3030e+01,  3.2431e+01,  ..., -9.7113e+01,
           -9.8713e+01, -9.3947e+01]],

         [[ 4.2641e+01,  4.4395e+01,  4.2862e+01,  ..., -1.1494e+01,
           -1.2772e+01, -1.4538e+01],
          [ 4.4528e+01,  4.5528e+01,  4.6945e+01,  ...,  6.1947e+00,
            2.2182e+00,  7.2983e+00],
          [ 4.7423e+01,  4.8762e+01,  4.4562e+01,  ...,  1.1089e+01,
            7.8785e+00,  9.8004e+00],
          ...,
          [ 2.3506e+01,  2.3662e+01,  2.6562e+01,  ..., -1.0194e+02,
           -1.0169e+02, -1.0094e+02],
          [ 2.0758e+01,  2.3419e+01,  2.8762e+01,  ..., -1.0359e+02,
           -1.0787e+02, -1.0177e+02],
          [ 2.6979e+01,  2.6205e+01,  2.5606e+01,  ..., -1.0394e+02,
           -1.0554e+02, -1.0077e+02]]]], device='cuda:0') tensor([[ 600.0000,  600.0000,    0.1172]], device='cuda:0') tensor([[[ 449.4141,  419.8828,  456.4453,  435.8203,    1.0000],
         [ 313.1250,  313.3594,  328.8281,  320.6250,    1.0000],
         [ 360.4688,  336.4453,  367.7344,  352.3828,    1.0000],
         ...,
         [   0.0000,    0.0000,    0.0000,    0.0000,    0.0000],
         [   0.0000,    0.0000,    0.0000,    0.0000,    0.0000],
         [   0.0000,    0.0000,    0.0000,    0.0000,    0.0000]]], device='cuda:0') tensor([ 380], device='cuda:0')
			rpn_cls: 0.0245, rpn_box: 0.1481, RCNN_cls: 0.2863, RCNN_box 0.1674
			src_img_loss: 7.3048, src_inst_loss: 1.7190, src_consistency_loss: 0.2121
			tar_img_loss: 0.0170, tar_inst_loss: 0.3771, tar_consistency_loss: 0.1855
WARNING: unstable losses!
tensor([[[[ 6.2253e+01,  6.2047e+01,  6.3653e+01,  ...,  8.3259e+00,
            8.0899e+00,  5.7866e+00],
          [ 6.6320e+01,  6.7110e+01,  7.0003e+01,  ...,  2.6620e+01,
            2.2530e+01,  2.6720e+01],
          [ 6.3981e+01,  6.2720e+01,  6.3853e+01,  ...,  3.1020e+01,
            2.4787e+01,  3.0553e+01],
          ...,
          [ 1.7492e+01,  1.8720e+01,  2.6853e+01,  ..., -8.4980e+01,
           -7.4280e+01, -7.6980e+01],
          [ 1.6250e+01,  1.8840e+01,  2.8837e+01,  ..., -7.8113e+01,
           -8.4171e+01, -8.5680e+01],
          [ 1.8199e+01,  2.2460e+01,  2.4253e+01,  ..., -7.6775e+01,
           -7.8610e+01, -7.5047e+01]],

         [[ 4.9287e+01,  4.9080e+01,  5.0687e+01,  ..., -4.6405e+00,
           -4.8765e+00, -7.1798e+00],
          [ 5.3354e+01,  5.4144e+01,  5.7037e+01,  ...,  1.3653e+01,
            9.5634e+00,  1.3754e+01],
          [ 5.1015e+01,  4.9754e+01,  5.0887e+01,  ...,  1.8053e+01,
            1.1820e+01,  1.7587e+01],
          ...,
          [ 4.5260e+00,  5.7535e+00,  1.3887e+01,  ..., -9.7947e+01,
           -8.7246e+01, -8.9947e+01],
          [ 3.2834e+00,  5.8738e+00,  1.5871e+01,  ..., -9.1080e+01,
           -9.7137e+01, -9.8647e+01],
          [ 5.2324e+00,  9.4933e+00,  1.1287e+01,  ..., -8.9741e+01,
           -9.1577e+01, -8.8013e+01]],

         [[ 4.2462e+01,  4.2255e+01,  4.3862e+01,  ..., -1.1466e+01,
           -1.1702e+01, -1.4005e+01],
          [ 4.6528e+01,  4.7318e+01,  5.0212e+01,  ...,  6.8283e+00,
            2.7382e+00,  6.9283e+00],
          [ 4.4189e+01,  4.2928e+01,  4.4062e+01,  ...,  1.1228e+01,
            4.9952e+00,  1.0762e+01],
          ...,
          [-2.2992e+00, -1.0717e+00,  7.0618e+00,  ..., -1.0477e+02,
           -9.4071e+01, -9.6772e+01],
          [-3.5418e+00, -9.5139e-01,  9.0455e+00,  ..., -9.7905e+01,
           -1.0396e+02, -1.0547e+02],
          [-1.5928e+00,  2.6681e+00,  4.4617e+00,  ..., -9.6566e+01,
           -9.8402e+01, -9.4838e+01]]]], device='cuda:0') tensor([[ 600.0000,  600.0000,    0.1172]], device='cuda:0') tensor([[[ 393.6328,  299.1797,  400.6641,  306.2109,    1.0000],
         [ 384.7266,  255.8203,  391.7578,  271.7578,    1.0000],
         [ 340.7812,  352.2656,  348.9844,  367.5000,    1.0000],
         ...,
         [   0.0000,    0.0000,    0.0000,    0.0000,    0.0000],
         [   0.0000,    0.0000,    0.0000,    0.0000,    0.0000],
         [   0.0000,    0.0000,    0.0000,    0.0000,    0.0000]]], device='cuda:0') tensor([ 423], device='cuda:0')
```

**Comment by MaddyThakker** [comment]
I also tried to reduce the size of input images to (512, 512), now the loss becomes nan after training for 10 iterations.

**Comment by MaddyThakker** [comment]
I solved the issue by training a Faster-RCNN Model on the source dataset and then using that saved model to resume training on the Domain-Adaptive Faster-RCNN Model.

**Comment by ReasonablyAnon** [comment]
Fuck You @MaddyThakker  - Your solution doesn't work.
